### PR TITLE
WIP - consume groups from identity providers during login 

### PR DESCRIPTION
**Pull request by liggitt**
POC for https://trello.com/c/8olzeZH3

This PR: 

1. Plumbs the identity name and identity provider group information in extra fields in the user.Info returned during login during an OAuth flow
2. Plumbs user groups and extra data through the cookie-based session during the login flow
3. Adds the ability to consume group membership from three identity providers:
    * RequestHeaderIdentityProvider, based on configurable headers
    * OpenIDIdentityProvider, based on configurable claims which may contain a string or string array
    * BasicAuthPasswordIdentityProvider, which may return a field named `groups` containing an array of strings
4. Persists the identity and identity provider groups on OAuthAuthorizeToken and OAuthAccessToken objects
5. Includes identity provider groups from the OAuthAccessToken object in the user's effective group memberships when authenticating with an OAuth token

TODO:
- [ ] Determine valid group names for identity provider groups (same restrictions as Group API objects?)
- [ ] Determine how to handle/guard against cookie overflow for excessive numbers of groups
- [ ] Unit tests for each IDP group consumption
- [ ] Integration tests for code and implicit OAuth flows

### CODE REVIEW

**Comment by liggitt** [comment]
cc @enj @deads2k @openshift/security

**Comment by iankko** [comment]
@liggitt 

Thank you for the PoC, Jordan!

Tested your proposal with OpenID provider (the other two to be tested later) locally against a Keycloak server with groups information federated from LDAP server, and it seems to be parsing optional group info claim correctly.

```
I0712 07:53:03.685443   25259 handler.go:160] Got auth data
I0712 07:53:03.705588   25259 kubelet.go:1889] SyncLoop (housekeeping)
I0712 07:53:03.756860   25259 openid.go:224] identity=&{rh_sso 53d5d7b6-9614-425c-ab93-afaef8338ed0 [ldapuser1 ldapuser2] map[preferred_username:iankko name:Jan Lieskovsky ]}
I0712 07:53:03.758217   25259 wrap.go:42] GET /oauth2callback/rh_sso?state=Y3NyZj1kZTgyYjk3ZC02NmUwLTExZTctODk3My01MjU0MDAyYjA0MTAmdGhlbj0lMkZvYXV0aCUyRmF1dGhvcml6ZSUzRmNsaWVudF9pZCUzRG9wZW5zaGlmdC13ZWItY29uc29sZSUyNmlkcCUzRHJoX3NzbyUyNnJlZGlyZWN0X3VyaSUzRGh0dHBzJTI1M0ElMjUyRiUyNTJGMTkyLjE2OC4xMDMuMTkyJTI1M0E4NDQzJTI1MkZjb25zb2xlJTI1MkZvYXV0aCUyNnJlc3BvbnNlX3R5cGUlM0Rjb2RlJTI2c3RhdGUlM0RleUowYUdWdUlqb2lMeUlzSW01dmJtTmxJam9pTVRRNU9UZzJNRE00TWpReE55MHhPREkzT0RFek1UYzBNakV4TVRnek56a3lNakl5TkRJeU5qQXlOelU0TnpjME1UWXhOVFF5TVRNNE5UTTRNREE1TkRFMk5UWTFPVFE0TVRVNE5qazJOalkyTnpNMU9UZ3hNREl5TkRVaWZR&code=Aj_V_T9ferpjT-oIovEGjSvKSgNhwXu4eRZJx1EcvVI.2af942fa-fdd6-4279-9eda-06b54c63ed3a: (72.844018ms) 302 [[Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:53.0) Gecko/20100101 Firefox/53.0] 192.168.103.1:55724]
I0712 07:53:03.766128   25259 authenticator.go:38] OAuth authentication succeeded: &user.DefaultInfo{Name:"iankko", UID:"57fcd708-66f5-11e7-97bb-5254002b0410", Groups:[]string(nil), Extra:map[string][]string{"name.identity.openshift.io":[]string{"rh_sso:53d5d7b6-9614-425c-ab93-afaef8338ed0"}, "groups.identity.openshift.io":[]string{"ldapuser1", "ldapuser2"}}}
```

(Btw. is there so far other way how to get this added group information for the user from their identity other than inspecting the logs of the master? It doesn't seem to be printed when describing the user or its identity:

```
# oc describe user iankko
Name:		iankko
Created:	About an hour ago
Labels:		<none>
Annotations:	<none>
Full Name:	Jan Lieskovsky 
Identities:	rh_sso:53d5d7b6-9614-425c-ab93-afaef8338ed0
```

```
# oc describe identity rh_sso:53d5d7b6-9614-425c-ab93-afaef8338ed0
Name:		rh_sso:53d5d7b6-9614-425c-ab93-afaef8338ed0
Created:	About an hour ago
Labels:		<none>
Annotations:	<none>
User Name:	iankko
User UID:	57fcd708-66f5-11e7-97bb-5254002b0410
```
)

Yet, while this change extracts group information from the claim (and adds it into user's identity), it doesn't create the groups themselves:

```
# oc get groups
No resources found.
# oc whoami
iankko
```

Is this expected? (IOW to be enhanced yet in the future)

Thanks, Jan

File: pkg/auth/oauth/registry/tokenauthenticator.go [hunk 1]
```diff
@@ -52,20 +53,25 @@ func (a *TokenAuthenticator) AuthenticateToken(value string) (kuser.Info, bool,
...
+	groupNames.Insert(u.Groups...)
+	// 3. Group API objects the user is listed as a member in
 	groups, err := a.groupMapper.GroupsFor(u.Name)
 	if err != nil {
 		return nil, false, err
 	}
-	groupNames := []string{}
 	for _, group := range groups {
-		groupNames = append(groupNames, group.Name)
+		groupNames.Insert(group.Name)
```
Comment (iankko) [inline pkg/auth/oauth/registry/tokenauthenticator.go]: Should this check if group with specific name isn't present in the set already? And insert it, only if not present yet -- something like in https://github.com/openshift/origin/blob/master/pkg/cmd/admin/groups/new.go#L127-L135

Comment (liggitt) [inline pkg/auth/oauth/registry/tokenauthenticator.go]: string set dedupes

**Comment by liggitt** [comment]
> is there so far other way how to get this added group information for the user from their identity other than inspecting the logs of the master?

no. audit logging is the best way currently:
```yaml
auditConfig:
  auditFilePath: path/to/audit.log
  enabled: true
  ...
```

> It doesn't seem to be printed when describing the user or its identity

correct, the group info obtained during login is associated with the session and OpenShift API token

> Yet, while this change extracts group information from the claim (and adds it into user's identity), it doesn't create the groups themselves:
> 
> # oc get groups
> No resources found.
> # oc whoami
> iankko
> 
> Is this expected? (IOW to be enhanced yet in the future)

Yes, it is expected. I think group membership obtained at login time from a particular identity provider should be associated with the session (e.g., that API token), rather than persisted into a Group API object, for several reasons:
* avoids write contention on Group API objects (if you have a group with thousands of users, every new login by any of those users would rewrite that group)
* avoids flutter if a user has two active API tokens obtained by authenticating via two different identity providers with different views of group membership
* avoids scanning all Group objects and scrubbing a user from them on every login
* avoids needing to expand the Group API object schema to track the source of all member items to know if the absence of a group name from the identity provider on login should result in a user being removed from a particular Group API object.
* allows cluster admins to continue to use locally-defined groups
* allows an external identity provider that is doing federated authentication to provide different group membership depending on login method, and have OpenShift honor that

**Comment by chrisruffalo** [comment]
I've just finished testing this and I feel like this is a good start. I'd like a bit more ability to inspect what groups have been assigned in this way (view groups on the session).

**Comment by iankko** [comment]
@chrisruffalo 
> I've just finished testing this and I feel like this is a good start. I'd like a bit more ability to inspect what groups have been assigned in this way (view groups on the session).

There's [enhancement to show the IdP groups possibly present in the token when describing the OAuthAccessToken](https://github.com/iankko/origin/commit/a22b6b762570f186cf722d234650e2138273f572).

If you apply that patch on top of Jordan's changes, the new `Identity Provider Groups` row will be shown when describing the OAuthAccessToken (one IdP groups displayed per line. For providers not having them, it will be ```<none>```). Thus something like:

```
# oc get oauthaccesstokens
NAME                                          USER NAME   CLIENT NAME                    CREATED                         EXPIRES                         REDIRECT URI                                        SCOPES
OIkQy4NE__keyI0hnhNwavq2j0b_zCeQTbL2wmWiTJs   iankko      openshift-challenging-client   2017-07-17 08:52:38 -0400 EDT   2017-07-18 08:52:38 -0400 EDT   https://192.168.103.192:8443/oauth/token/implicit   user:full
eEz6EfYrDucwFQTWiXJik6ww3uQsMSM5xos3VNb0xxg   iankko      openshift-challenging-client   2017-07-19 04:34:54 -0400 EDT   2017-07-20 04:34:54 -0400 EDT   https://192.168.103.192:8443/oauth/token/implicit   user:full

# oc describe oauthaccesstoken eEz6EfYrDucwFQTWiXJik6ww3uQsMSM5xos3VNb0xxg
Name:				eEz6EfYrDucwFQTWiXJik6ww3uQsMSM5xos3VNb0xxg
Created:			About a minute ago
Labels:				<none>
Annotations:			<none>
Scopes:				[user:full]
Expires In:			24 hours
User Name:			iankko
User UID:			72119adb-6aee-11e7-8752-5254002b0410
Client Name:			openshift-challenging-client
Identity Provider Groups:	ldapuser1
				ldapuser2
 ```
Note: This will obviously work only for user(s) having sufficient privileges (privilege to display oauthaccesstokens). We are [exploring yet if the very same information could be made accessible to regular user too in some way](https://trello.com/c/8olzeZH3/34-security-as-a-cluster-administrator-i-can-configure-openshift-to-consume-group-membership-information-from-an-identity-provider#comment-596cff1dadbc5494e9b7aa3d).

**Comment by liggitt** [comment]
a describer on the token still doesn't really do what the user wants, which is "tell me the effective groups I have". those groups can come from the User object (deprecated, but still present for now), the token (with this PR), the Group API objects, virtual groups (system:authenticated, etc), etc. Surfacing the effective group membership would be a new API (one I'm in favor of, but separate from this)

**Comment by chrisruffalo** [comment]
@iankko That worked perfectly.

@liggitt From a consulting and ops perspective I'm not particularly worried about the user at this time. I'd like to be able to expose that to users eventually but this feature is huge, to me, in driving adoption of other forms of authentication and changing the way that adopters think about group-sync and IdM for OCP.

Thank you both for your help in testing this one.

**Comment by smarterclayton** [comment]
Want

On Thu, Sep 7, 2017 at 4:17 PM, OpenShift CI Robot <notifications@github.com

**Comment by enj** [comment]
Moving this to 3.8 since it will not make it for 3.7.

**Comment by simo5** [comment]
@JohannesRudolph work has been resumed here #20262

**Pull request closed without merging**